### PR TITLE
Update wawaka support for WAMR-03-05-2020

### PR DESCRIPTION
**Pull request by cmickeyb**
Update the wawaka to support WAMR tag release WAMR-03-05-2020. Several
changes to the build process and to the integration of native functions.
There are still opportunities to take advantage of the parameter passing
for pointers, however I think it best to keep the validation of pointers
that we have.

Must get the correct submodule for WAMR.

Signed-off-by: Mic Bowman <mic.bowman@intel.com>

### CODE REVIEW

**Comment by cmickeyb** [comment]
@masomel i think our handling of the submodules in general is still broken. either we need to fix the documentation to reflect how we believe the cloning should work or we need to make sure it works as documented (which i do not believe it does)

File: build/common-config.sh [hunk 1]
```diff
@@ -35,7 +35,7 @@ var_set() {
 	"
 	env_key_sort[$i]="TINY_SCHEME_SRC"; i=$i+1; export TINY_SCHEME_SRC=${env_val[TINY_SCHEME_SRC]};
 
-	env_val[WASM_SRC]="${WASM_SRC:-${PDO_SOURCE_ROOT}/wasm}"
+	env_val[WASM_SRC]="${WASM_SRC:-${PDO_SOURCE_ROOT}/interpreters/wasm-micro-runtime}"
 	env_desc[WASM_SRC]="
```
Comment (prakashngit) [inline build/common-config.sh]: Given that this is getting checked out every time to ${PDO_SOURCE_ROOT}/interpreters/wasm-micro-runtime}, bash variable for WASM_SRC might be confusing. May I suggest that we hard code WASM_SRC to ${PDO_SOURCE_ROOT}/interpreters/wasm-micro-runtime}?

Comment (masomel) [reply build/common-config.sh]: By hard code the value of `WASM_SRC`, you mean, remove the `env_val[WASM_SRC]=` portion?

**Comment by masomel** [comment]
> @masomel i think our handling of the submodules in general is still broken. either we need to fix the documentation to reflect how we believe the cloning should work or we need to make sure it works as documented (which i do not believe it does)

I'm still a bit unclear as to the exact behavior your want from the submodule. The git command to clone the PDO repo including all submodules seems to be working for folks. We certainly need to update the documentation, in any case. I've been working to fix the Travis issue, it's not related to how the submodule is handled. The docker config file is outdated.

File: build/common-config.sh (hunk 1 above)
Comment (prakashngit) [reply build/common-config.sh]: avoid the optional WASM_SRC bash variable. not a big deal (I had to unset my variable for building this current PR)

**Comment by cmickeyb** [comment]
> > @masomel i think our handling of the submodules in general is still broken. either we need to fix the documentation to reflect how we believe the cloning should work or we need to make sure it works as documented (which i do not believe it does)
> 
> I'm still a bit unclear as to the exact behavior your want from the submodule. The git command to clone the PDO repo including all submodules seems to be working for folks. We certainly need to update the documentation, in any case. I've been working to fix the Travis issue, it's not related to how the submodule is handled. The docker config file is outdated.

I want the instructions to work on a clean, pristine repository. That is not currently the case. We have no reference I can find anywhere that the tagged version should be pulled. We are currently "lucky" that head seems to be working.

**Comment by masomel** [comment]
> I want the instructions to work on a clean, pristine repository. That is not currently the case. We have no reference I can find anywhere that the tagged version should be pulled. We are currently "lucky" that head seems to be working.

So this sounds like this can be solved by updating the build instructions. We shouldn't need to add extra instructions to pull in the tagged version in a pristine clone as long as (1) the submodule points to the right tagged commit (2) people clone the PDO repo with the `--recurse-submodules` flag.

**Comment by masomel** [comment]
> Looks like we need to make changes to https://github.com/hyperledger-labs/private-data-objects/blob/master/docker/Dockerfile.pdo-dev#L189 and beyond to make Travis happy.

As long as Travis is using the docker tests, it's going to fail, since  the Dockerfile clones directly master, which does _not_ currently include the WAMR submodule. I will change this file temporarily for testing purposes, but we can't merge the PR with a Dockerfile that points to a fork of this repo.

File: build/common-config.sh (hunk 1 above)
Comment (cmickeyb) [reply build/common-config.sh]: i would prefer to keep it. it still allows for experimentation with alternate branches/tags.

File: common/interpreter/wawaka_wasm/WasmExtensions.cpp [hunk 2]
```diff
@@ -37,70 +37,16 @@
 
 namespace pe = pdo::error;
 
-/* ----------------------------------------------------------------- *
```
Comment (g2flyer) [inline common/interpreter/wawaka_wasm/WasmExtensions.cpp]: Did we had to change order of method declaration?  Makes diff pretty useless ... :(

File: docker/Dockerfile.pdo-build [hunk 3]
```diff
@@ -101,6 +102,13 @@ ENV PDO_DEBUG_BUILD=${PDO_DEBUG_BUILD}
 ARG PDO_INTERPRETER=gipsy
 ENV PDO_INTERPRETER=${PDO_INTERPRETER}
 
+# - web-assembly/wasm/wawaka
+#   - Configure WAMR source
+ARG WAMR=WAMR-03-05-2020
+RUN cd /project/pdo/src/private-data-objects/interpreters/wasm-micro-runtime \
+&& git checkout ${WAMR}
```
Comment (g2flyer) [inline docker/Dockerfile.pdo-build]: a trailing \ is missing (and for stylistic reasons might be nice to preceed && with some spaces?)

File: common/interpreter/wawaka_wasm/WasmExtensions.cpp (hunk 2 above)
Comment (cmickeyb) [reply common/interpreter/wawaka_wasm/WasmExtensions.cpp]: yes. there are a number of functions that were just removed. pretty much everything else had to change.

**Pull request merged**